File: code/sorting/circle_sort/circle_sort.m
/* Part of Cosmos by OpenGenus Foundation */

//
//  circle_sort.m
//  Created by DaiPei on 2017/10/15.
//

#import <Foundation/Foundation.h>

@interface CircleSort : NSObject

- (void)sort:(NSMutableArray<NSNumber *> *)array;

@end

@implementation CircleSort

- (void)sort:(NSMutableArray<NSNumber *> *)array {
    while ([self sort:array low:0 high:(int)array.count - 1]) {}
}

- (BOOL)sort:(NSMutableArray<NSNumber *> *)array low:(int)low high:(int)high {
    BOOL swapped = NO;
    if (low == high) {
        return swapped;
    }
    int l = low, h = high;
    while (l < h) {
        if ([array[l] compare:array[h]] == NSOrderedDescending) {
            [self swap:array at:l and:h];
            swapped = YES;
        }
        l++;
        h--;
    }
    if (l == h) {
        if ([array[l] compare:array[l + 1]] == NSOrderedDescending) {
            [self swap:array at:l and:l + 1];
            swapped = YES;
        }
    }
    int mid = (low + high) / 2;
    BOOL left = [self sort:array low:low high:mid];
    BOOL right = [self sort:array low:mid + 1 high:high];
    
    return swapped || left || right;
}

- (void)swap:(NSMutableArray<NSNumber *> *)array at:(NSUInteger)indexA and:(NSUInteger)indexB {
    NSNumber *tmp = array[indexA];
    array[indexA] = array[indexB];
    array[indexB] = tmp;
}

@end

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSMutableArray *array = [NSMutableArray arrayWithCapacity:10];
        for (int i = 0; i < 30; i++) {
            int ran = arc4random() % 100 - 50;
            [array addObject:@(ran)];
        }
        NSLog(@"before: %@", array);
        CircleSort *cs = [[CircleSort alloc] init];
        [cs sort:array];
        NSLog(@"after: %@", array);
    }
    return 0;
}

File: code/sorting/circle_sort/circle_sort.swift
/* Part of Cosmos by OpenGenus Foundation */

//
//  circle_sort.swift
//  Created by DaiPei on 2017/10/14.
//

import Foundation

func circleSort(_ array: inout [Int]) {
    while circleSort(&array, low: 0, high: array.count - 1) {}
}

private func circleSort(_ array: inout [Int], low: Int, high: Int) -> Bool {
    var swaped = false
    if low == high {
        return swaped
    }
    var l = low
    var h = high
    while l < h {
        if array[l] > array[h] {
            swap(&array, at: l, and: h)
            swaped = true
        }
        l += 1
        h -= 1
    }
    if l == h {
        if l + 1 < array.count && array[l] > array[l + 1] {
            swap(&array, at: l, and: l + 1)
            swaped = true
        }
    }
    let mid = (low + high) / 2
    let left = circleSort(&array, low: low, high: mid)
    let right = circleSort(&array, low: mid + 1, high: high)
    return swaped || left || right
}

private func swap(_ array: inout [Int], at indexA: Int, and indexB: Int) {
    let tmp = array[indexA]
    array[indexA] = array[indexB]
    array[indexB] = tmp
}
